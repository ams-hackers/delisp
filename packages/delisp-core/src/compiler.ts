import { ASExpr, ASExprSymbol, ASExprList } from "./syntax";
import * as recast from "recast";

const debug = require("debug")("delisp:compiler");

type JSAST = object;
interface Environment {
  [symbol: string]: { symbol: ASExprSymbol; jsVar: string };
}

function parseLambdaList(x: ASExpr): ASExprSymbol[] {
  if (x.type !== "list")
    throw new Error("First argument of lambda should be a list");

  if (!x.elements.every(param => param.type === "symbol"))
    throw new Error("First argument of lambda should be a list of symbols");

  return x.elements as ASExprSymbol[];
}

function compileLambda(args: ASExpr[], env: Environment): JSAST {
  if (args.length !== 2) throw new Error("Lambda needs exactly 2 arguments");

  const params = parseLambdaList(args[0]);

  const newEnv = params.reduce(
    (e, param, ix) => ({
      ...e,
      [param.name]: { symbol: param, jsVar: `p${ix}` }
    }),
    env
  );

  return {
    type: "ArrowFunctionExpression",
    params: params.map(symbol => ({
      type: "Identifier",
      name: newEnv[symbol.name].jsVar
    })),
    body: compile(args[1], newEnv)
  };
}

function compileDefinition(args: ASExpr[], env: Environment): JSAST {
  if (args.length !== 2) throw new Error("Define needs exactly 2 arguments");

  const variable = args[0];
  if (variable.type !== "symbol")
    throw new Error("First argument of define should be a symbol");

  return {
    type: "AssignmentExpression",
    operator: "=",
    left: {
      type: "MemberExpression",
      computed: true,
      object: {
        type: "Identifier",
        name: "env"
      },
      property: {
        type: "Literal",
        value: variable.name
      }
    },
    right: compile(args[1], env)
  };
}

function compileList(fn: ASExpr, args: ASExpr[], env: Environment): JSAST {
  if (fn.type === "symbol") {
    switch (fn.name) {
      case "lambda":
        return compileLambda(args, env);
    }
  }

  return {
    type: "CallExpression",
    callee: compile(fn, env),
    arguments: args.map(arg => compile(arg, env))
  };
}

function compileVariable(symbol: ASExprSymbol, env: Environment): JSAST {
  if (symbol.name in env) {
    return {
      type: "Identifier",
      name: env[symbol.name].jsVar
    };
  } else {
    return {
      type: "MemberExpression",
      computed: true,
      object: {
        type: "Identifier",
        name: "env"
      },
      property: {
        type: "Literal",
        value: symbol.name
      }
    };
  }
}

export function compile(syntax: ASExpr, env: Environment): JSAST {
  switch (syntax.type) {
    case "number":
      return {
        type: "NumericLiteral",
        value: syntax.value
      };
    case "symbol":
      return compileVariable(syntax, env);
    case "string":
      return {
        type: "Literal",
        value: syntax.value
      };
    case "list":
      const [fn, ...args] = syntax.elements;
      return compileList(fn, args, env);
  }
}

function isSpecialForm(syntax: ASExprList, name: string) {
  if (syntax.elements.length > 0) {
    const first = syntax.elements[0];
    return first.type === "symbol" && first.name === name;
  } else {
    return false;
  }
}

function compileTopLevel(syntax: ASExpr, env: Environment): JSAST {
  if (syntax.type === "list" && isSpecialForm(syntax, "define")) {
    return compileDefinition(syntax.elements.slice(1), env);
  } else {
    return compile(syntax, env);
  }
}

export function compileModule(syntax: ASExpr): JSAST {
  return {
    type: "File",
    program: {
      type: "Program",
      sourceType: "module",
      body: [
        {
          type: "ExpressionStatement",
          expression: compileTopLevel(syntax, {})
        }
      ]
    }
  };
}

export function compileToString(syntax: ASExpr): string {
  const ast = compileModule(syntax);
  debug("jsast:", ast);
  return recast.print(ast).code;
}
