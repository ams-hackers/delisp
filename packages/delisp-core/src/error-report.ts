/**
 * Report errors in a user-friendly way
 */

import { Location } from "./input";

function repeatChar(ch: string, n: number): string {
  return Array(n)
    .fill(ch)
    .join("");
}

/** Print a error message with some highlighted piece of a source code */
export function printHighlightedSource(
  message: string,
  source: string,
  offset: number
) {
  const lines = source.split("\n");

  // Calculate the `line` and `column` (0 based) for this offset in
  // source.
  let line = 0;
  let remainingOffset = offset;
  while (lines.length > line && remainingOffset >= lines[line].length + 1) {
    remainingOffset -= lines[line].length;
    line++;
  }
  const column = remainingOffset;

  return [
    `file:${line + 1}:${column}: ${message}`,
    lines[line],
    repeatChar("-", column) + "^"
  ].join("\n");
}

/** Print a error message with expr highlighted. */
export function printHighlightedExpr(
  message: string,
  location: Location,
  end = false
) {
  const source = location.input.toString();
  const offset = end ? location.end - 1 : location.start;
  return printHighlightedSource(message, source, offset);
}
