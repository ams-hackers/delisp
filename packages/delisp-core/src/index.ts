export * from "./compiler";
export * from "./syntax-convert";
export * from "./error-report";
export * from "./eval";
export * from "./infer";
export * from "./infer-environment";
export * from "./module";
export * from "./module-dependencies";
export * from "./prettier";
export * from "./primitives";
export * from "./printer";
export * from "./syntax";
export * from "./syntax-typed";
export * from "./syntax-scoped";
export * from "./syntax-utils";
export * from "./type-printer";
export * from "./type-utils";
export { Type } from "./types";
export * as T from "./types";
export * from "./typescript-generation";
export * from "./macroexpand";
