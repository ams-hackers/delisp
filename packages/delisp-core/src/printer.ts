import { assertNever, InvariantViolation } from "./invariant";
import {
  align,
  concat,
  Doc,
  encodeMany,
  Encoder,
  group,
  groupalign,
  indent as indent_,
  join,
  line,
  nil,
  prettyAs,
  space,
  StringEncoder,
  text
} from "./prettier";
import * as S from "./syntax";
import { foldExpr } from "./syntax-utils";
import { printType } from "./type-printer";

function indent(x: Doc, level = 2): Doc {
  return indent_(x, level);
}

function printString(str: string, source: S.Syntax): Doc {
  const escaped = str.replace(/\n/g, "\\n").replace(/"/g, '\\"');
  return text(`"${escaped}"`, ["string"], source);
}

function printIdentifier(
  name: string,
  source?: S.Syntax,
  kinds: string[] = []
): Doc {
  return text(name, ["identifier", ...kinds], source);
}

function keyword(name: string, source?: S.Syntax): Doc {
  return text(name, ["keyword"], source);
}

function list(...docs: Doc[]): Doc {
  return concat(text("(", ["delimiter"]), ...docs, text(")", ["delimiter"]));
}

function vector(...docs: Doc[]): Doc {
  return concat(text("[", ["delimiter"]), ...docs, text("]", ["delimiter"]));
}

function map(...docs: Doc[]): Doc {
  return concat(text("{", ["delimiter"]), ...docs, text("}", ["delimiter"]));
}

function lines(...docs: Doc[]): Doc {
  return concat(line, join(docs, line));
}

function printCase(pattern: Doc, body: Doc[]) {
  return group(concat(list(pattern, indent(lines(...body)))));
}

function printExpr(expr: S.Expression): Doc {
  return foldExpr(expr, (e, source) => {
    switch (e.node.tag) {
      case "unknown": {
        if (!e.location) {
          throw new InvariantViolation(
            `Unknown syntax node has no location information.`
          );
        }
        const { input, start, end } = e.location;
        return text(input.toString().slice(start, end), ["unknown"]);
      }
      case "string":
        return printString(e.node.value, source);
      case "number":
        return text(String(e.node.value), ["number"], source);
      case "vector":
        return group(vector(align(...e.node.values)));
      case "boolean":
        return text(e.node.value.toString(), ["boolean"], source);
      case "record": {
        return group(
          map(
            align(
              join(
                [
                  ...e.node.fields.map(({ label, value }) =>
                    concat(text(label.name, ["label"]), space, value)
                  ),
                  ...(e.node.source
                    ? [concat(text("|", ["label-ext"]), space, e.node.source)]
                    : [])
                ],
                line
              )
            )
          )
        );
      }
      case "record-get":
        return group(
          list(
            concat(
              keyword("$get", source),
              space,
              text(e.node.field.name, ["label"], e),
              space,
              e.node.value
            )
          )
        );
      case "variable-reference":
        return printIdentifier(e.node.name, { ...e, node: e.node });

      case "conditional":
        return group(
          list(
            concat(
              keyword("if", source),
              space,
              align(e.node.condition, e.node.consequent, e.node.alternative)
            )
          ),
          // NOTE: We don't want conditionals to be too long. We use a
          // much lower value here, we'll only group them if they are
          // realluy short.
          30
        );

      case "function":
        const singleBody = e.node.body.length === 1;
        const doc = list(
          keyword("lambda", source),
          space,
          group(
            list(
              align(
                ...e.node.lambdaList.userPositionalArguments.map(x =>
                  printIdentifier(x.name, undefined, ["argument"])
                )
              )
            )
          ),
          indent(lines(...e.node.body))
        );
        return singleBody ? group(doc) : doc;

      case "function-call": {
        const fn = e.node.fn;
        const args = e.node.userArguments;
        return group(list(groupalign(fn, align(...args))));
      }

      case "let-bindings":
        return list(
          keyword("let", source),
          space,
          map(
            align(
              ...e.node.bindings.map(b =>
                concat(
                  printIdentifier(b.variable.name, undefined, [
                    "lexical-variable-declaration"
                  ]),
                  space,
                  b.value
                )
              )
            )
          ),
          indent(lines(...e.node.body))
        );

      case "type-annotation":
        return group(
          list(
            keyword("the", source),
            space,
            text(e.node.typeWithWildcards.print(), ["type"]),
            indent(concat(line, e.node.value))
          )
        );

      case "do-block":
        return list(
          concat(
            keyword("do", source),
            indent(lines(...e.node.body, e.node.returning))
          )
        );

      case "match": {
        return list(
          concat(
            keyword("match", source),
            space,
            e.node.value,
            indent(
              concat(
                line,
                join(
                  [
                    ...e.node.cases.map(c => {
                      return printCase(
                        map(
                          keyword(c.label),
                          space,
                          printIdentifier(c.variable.name)
                        ),
                        c.body
                      );
                    }),

                    e.node.defaultCase
                      ? printCase(keyword(":default"), e.node.defaultCase)
                      : nil
                  ],
                  line
                )
              )
            )
          )
        );
      }

      case "case":
        return group(
          list(
            keyword("case", source),
            space,
            text(e.node.label, ["label"]),
            e.node.value ? indent(concat(line, e.node.value)) : nil
          )
        );

      case "values": {
        return group(
          list(groupalign(keyword("values", source), align(...e.node.values)))
        );
      }

      case "multiple-value-bind":
        return list(
          keyword("multiple-value-bind", source),
          space,
          // List of variables
          list(
            join(
              e.node.variables.map(v =>
                printIdentifier(v.name, source, ["lexical-variable-definition"])
              ),
              space
            )
          ),
          indent(concat(line, e.node.form), 4),
          indent(lines(...e.node.body))
        );
    }
  });
}

function print(form: S.Syntax): Doc {
  if (S.isExpression(form)) {
    return printExpr(form);
  } else {
    switch (form.node.tag) {
      case "definition":
        return group(
          list(
            keyword("define", form),
            space,
            printIdentifier(form.node.variable.name, undefined, [
              "variable-definition"
            ]),
            indent(concat(line, print(form.node.value)))
          )
        );

      case "import":
        return list(
          keyword("import", form),
          space,
          printIdentifier(form.node.variable.name),
          space,
          keyword(":from"),
          space,
          printString(form.node.source, form)
        );

      case "export":
        return list(
          keyword("export", form),
          space,
          form.node.identifiers.length === 1
            ? printIdentifier(form.node.identifiers[0].name, form)
            : group(
                vector(
                  align(
                    ...form.node.identifiers.map(i =>
                      printIdentifier(i.name, form)
                    )
                  )
                )
              )
        );

      case "type-alias":
        return group(
          list(
            keyword("type", form),
            space,
            printIdentifier(form.node.alias.name, undefined, ["definition"]),
            indent(
              concat(
                line,
                text(printType(form.node.definition, false), ["type"])
              )
            )
          )
        );
      default:
        return assertNever(form.node);
    }
  }
}

export function pprintAs<A>(
  form: S.Syntax,
  lineWidth: number,
  encoder: Encoder<A>
): A {
  return prettyAs(print(form), lineWidth, encoder);
}

export function pprintModuleAs<A>(
  m: S.Module,
  lineWidth: number,
  encoder: Encoder<A>
): A {
  return encodeMany(
    encoder,
    m.body.map((s, i) => {
      let newlines: number;
      if (i > 0) {
        const previousForm = m.body[i - 1];

        const end = previousForm.location
          ? previousForm.location.end
          : undefined;
        const start = s.location && s.location.start;
        const between =
          s.location && s.location.input.toString().slice(end, start);
        newlines = between ? between.split("\n").length - 1 : 1;
      } else {
        newlines = 0;
      }

      const nl =
        newlines > 1
          ? encoder.fromString("\n", ["space"])
          : encoder.fromString("", []);
      return encoder.concat(nl, pprintAs(s, lineWidth, encoder));
    }),
    encoder.fromString("\n", ["space"])
  );
}

export function pprint(form: S.Syntax, lineWidth: number): string {
  return prettyAs(print(form), lineWidth, StringEncoder);
}

export function pprintModule(m: S.Module, lineWidth: number): string {
  return pprintModuleAs(m, lineWidth, StringEncoder);
}
