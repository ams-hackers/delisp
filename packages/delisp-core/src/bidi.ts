import { InvariantViolation, assertNever } from "./invariant";
import { Expression } from "./syntax";
import { Type, tString, tNumber } from "./types";
import { unify } from "./unify";
import { printType } from "./type-printer";
import { last } from "./utils";
import { Location } from "./input";
import { printHighlightedExpr } from "./error-report";

export interface Environment {
  [variable: string]: Type;
}

function equalTypes(t1: Type, t2: Type): boolean {
  // Quick and dirty equality definition for trying out.
  const result = unify(t1, t2, {});
  return (
    result.tag === "unify-success" &&
    Object.keys(result.substitution).length === 0
  );
}

function typeError(message: string, entity: { location: Location }) {
  return new Error(printHighlightedExpr(message, entity.location));
}

export function infer(expr: Expression, env: Environment): Type | null {
  switch (expr.node.tag) {
    case "string":
      return tString;
    case "number":
      return tNumber;
    case "variable-reference": {
      const variableName = expr.node.name;
      const variableType = env[variableName];
      if (!variableType) {
        throw typeError(`Unknown variable ${variableName}`, expr);
      }
      return variableType;
    }
    case "type-annotation": {
      const typ = expr.node.typeWithWildcards.instantiate();
      check(expr.node.value, typ, env);
      return typ;
    }
    case "function-call": {
      const fnType = infer(expr.node.fn, env);
      if (!fnType) {
        throw typeError(
          `Can't called a function with unknown type.`,
          expr.node.fn
        );
      }

      if (
        !(
          fnType.node.tag === "application" &&
          fnType.node.op.node.tag === "constant" &&
          fnType.node.op.node.name === "->"
        )
      ) {
        throw typeError(
          `Looks like you are trying to do a function call on a ${printType(
            fnType
          )}`,
          expr.node.fn
        );
      }

      const argTypes = fnType.node.args.slice(0, -1);
      const outType = last(fnType.node.args);
      if (!outType) {
        throw new InvariantViolation(`Function with no output type?`);
      }

      expr.node.args.forEach((arg, i) => {
        check(arg, argTypes[i], env);
      });

      return outType;
    }

    default:
      return null;
  }
}

export function check(expr: Expression, type: Type, env: Environment) {
  switch (expr.node.tag) {
    case "string":
    case "number":
    case "variable-reference":
    case "type-annotation":
    case "function-call": {
      const inferred = infer(expr, env);
      if (inferred) {
        if (!equalTypes(inferred, type)) {
          throw typeError(
            `Expected type ${printType(type, false)}, but found ${printType(
              inferred
            )}.`,
            expr
          );
        }
      }
      return;
    }
    case "function": {
      if (
        !(
          type.node.tag === "application" &&
          type.node.op.node.tag === "constant" &&
          type.node.op.node.name === "->"
        )
      ) {
        throw typeError(
          `Expected a function, but found ${printType(type)}`,
          expr
        );
      }

      const argTypes = type.node.args.slice(0, -1);
      const outType = last(type.node.args);
      if (!outType) {
        throw new InvariantViolation(`Function with no output type?`);
      }

      const args = expr.node.lambdaList.positionalArgs.map(i => i.name);

      if (argTypes.length !== args.length) {
        throw typeError(
          `Wrong number of arguments. Expected ${argTypes.length} but found ${
            args.length
          }`,
          expr.node.lambdaList
        );
      }

      const newEnv = args.reduce(
        (tmpEnv, a, idx) => ({
          ...tmpEnv,
          [a]: argTypes[idx]
        }),
        env
      );

      check(last(expr.node.body)!, outType, newEnv);
      return;
    }
    case "conditional":
    case "let-bindings":
    case "record":
    case "vector":
      throw typeError(
        `Cannot check this expression for type ${printType(type)}.`,
        expr
      );
    default:
      assertNever(expr.node);
  }
}
