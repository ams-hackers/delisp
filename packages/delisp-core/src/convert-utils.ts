import { printHighlightedExpr } from "./error-report";
import { ASExpr, ASExprMap, ASExprSymbol } from "./sexpr";
import { duplicatedItemsBy, last } from "./utils";

export class ConvertError extends Error {}

export interface ParseRecordResult {
  fields: Array<{
    label: ASExprSymbol;
    value: ASExpr;
  }>;
  tail: ASExpr | undefined;
}

export function parseRecord(expr: ASExprMap): ParseRecordResult {
  //
  // Destructure a map into fields and tail
  function fieldsAndTail() {
    if (expr.fields.length === 0) {
      return { fields: expr.fields };
    } else {
      const lastField = last(expr.fields)!;
      return lastField.label.name === "|"
        ? {
            fields: expr.fields.slice(0, -1),
            tail: lastField.value
          }
        : { fields: expr.fields };
    }
  }

  function checkInvalidFields() {
    const invalidFields = fields.filter(f => !f.label.name.startsWith(":"));
    if (invalidFields.length > 0) {
      const field = invalidFields[0];
      throw new ConvertError(
        printHighlightedExpr(
          `'${
            field.label.name
          }' is not a valid field name. Fields should start with ':' (colon).`,
          field.label.location
        )
      );
    }
  }

  const { fields, tail } = fieldsAndTail();
  checkInvalidFields();

  const duplicates = duplicatedItemsBy(fields, f => f.label.name);
  if (duplicates.length > 0) {
    throw new ConvertError(
      printHighlightedExpr("Duplicated label", duplicates[0].label.location)
    );
  }

  return { fields, tail };
}
