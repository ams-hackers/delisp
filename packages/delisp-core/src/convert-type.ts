import { parseRecord } from "./convert-utils";
import { printHighlightedExpr } from "./error-report";
import {
  ASExpr,
  ASExprList,
  ASExprMap,
  ASExprSymbol,
  ASExprVector
} from "./sexpr";
import { generateUniqueTVar } from "./type-generate";
import {
  emptyRow,
  Monotype,
  tApp,
  tBoolean,
  tNumber,
  tRecord,
  tString,
  tUserDefined,
  tVar,
  tVector,
  tVoid
} from "./types";

/** Capitalize a string like "foo" to "Foo". */
function capitalize(str: string) {
  if (str === "") {
    return "";
  } else {
    return str[0].toUpperCase() + str.slice(1);
  }
}

/** Return true if a symbol is a valid name for a user defined type, false otherwise. */
export function userDefinedType(expr: ASExprSymbol): boolean {
  return expr.name[0] === expr.name[0].toUpperCase();
}

/** Check if a symbol is a valid user defined type name or throw a user-friendly error otherwise. */
export function checkUserDefinedTypeName(expr: ASExprSymbol): void {
  if (!userDefinedType(expr)) {
    throw new Error(
      printHighlightedExpr(
        `'${
          expr.name
        }' is not a valid name for a type. Type names should start with a capital letter. Try '${capitalize(
          expr.name
        )}'?`,
        expr.location
      )
    );
  }
}

function convertSymbol(expr: ASExprSymbol): Monotype {
  switch (expr.name) {
    case "boolean":
      return tBoolean;
    case "number":
      return tNumber;
    case "string":
      return tString;
    case "void":
      return tVoid;
    case "_":
      return generateUniqueTVar(false, "__t");
    default:
      return userDefinedType(expr) ? tUserDefined(expr.name) : tVar(expr.name);
  }
}

function convertList(expr: ASExprList): Monotype {
  const [op, ...args] = expr.elements;

  if (op.type !== "symbol") {
    throw new Error(
      printHighlightedExpr("Expected symbol as operator", expr.location)
    );
  }

  switch (op.name) {
    case "->":
      if (args.length < 1) {
        throw new Error(
          printHighlightedExpr(
            "Expected at least 1 argument",
            op.location,
            true
          )
        );
      }
      break;
    default:
      throw new Error(
        printHighlightedExpr("Unknown type constructor", op.location)
      );
  }

  return tApp(op.name, ...args.map(convert));
}

function convertVector(expr: ASExprVector): Monotype {
  if (expr.elements.length !== 1) {
    throw new Error(
      printHighlightedExpr("Expected exactly 1 argument", expr.location)
    );
  }
  return tVector(convert(expr.elements[0]));
}

function convertMap(expr: ASExprMap): Monotype {
  const { fields, tail } = parseRecord(expr);

  return tRecord(
    fields.map(({ label, value }) => ({
      label: label.name,
      type: convert(value)
    })),
    tail ? convert(tail) : emptyRow
  );
}

/* Try to convert a S-Expression into a type. */
export function convert(expr: ASExpr): Monotype {
  switch (expr.type) {
    case "list":
      return convertList(expr);
    case "symbol":
      return convertSymbol(expr);
    case "vector":
      return convertVector(expr);
    case "map":
      return convertMap(expr);
    default:
      throw new Error(printHighlightedExpr("Not a valid type", expr.location));
  }
}
