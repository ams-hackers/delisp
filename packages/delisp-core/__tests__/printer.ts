import { readSyntax } from "../src/index";
import { pprint } from "../src/printer";

function pprintString(source: string, lineWidth: number = 80) {
  const syntax = readSyntax(source);
  return pprint(syntax, lineWidth);
}

describe("Pretty Printer", () => {
  it("should print lambda abstractions beautifully", () => {
    expect(pprintString("(lambda (aaa bbb ccc) xxx)")).toMatchSnapshot();
    expect(
      pprintString(
        "(lambda (aaa bbb ccc ddd eee fff ggg hhh iii jjj kkk lll mmm nnn ooo ppp qqq) xxx)"
      )
    ).toMatchSnapshot();
    expect(
      pprintString(
        "(lambda (aaa bbb ccc ddd eee fff ggg hhh iii jjj kkk lll mmm nnn ooo ppp qqq rrr sss ttt uuu vvv www xx) xxx)"
      )
    ).toMatchSnapshot();
  });
});
