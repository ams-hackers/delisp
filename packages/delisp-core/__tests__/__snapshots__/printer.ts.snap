// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pretty Printer should print lambda abstractions beautifully 1`] = `"(lambda (aaa bbb ccc) xxx)"`;

exports[`Pretty Printer should print lambda abstractions beautifully 2`] = `
"(lambda (aaa bbb ccc ddd eee fff ggg hhh iii jjj kkk lll mmm nnn ooo ppp qqq)
  xxx)"
`;

exports[`Pretty Printer should print lambda abstractions beautifully 3`] = `
"(lambda (aaa
         bbb
         ccc
         ddd
         eee
         fff
         ggg
         hhh
         iii
         jjj
         kkk
         lll
         mmm
         nnn
         ooo
         ppp
         qqq
         rrr
         sss
         ttt
         uuu
         vvv
         www
         xx)
  xxx)"
`;
